### ISSUE TITLE
Attributes for <script> tags?

### ISSUE THREAD

**Issue by tisba** [haml/haml#934]
I'm currently looking into enabling CSP ([1] via [2]) and I've some places where I cannot remove `<script>` directly. In CSP you can use a `nonce` as an attribute for `<script>` tags to allow inline evaluation of JavaScript.

How can I pass tag attributes with `:javascript` so that they will be rendered to the `<script>`-tag?

Looking at the current implementation it seems as if `Haml::Filters::Javascript.render_with_options` is not designed to take attributes. Is this intentional? I wasn't able to fully follow how `#render_with_options` is invoked and where it's `options` parameter comes from, but I'm happy to provide a PR if someone could point me in the right direction.

Ideally I'd like to write something like this (note that `content_security_policy_script_nonce` is a view helper from the `secure_headers` gem:

```haml
#bla
  :javascript noince: content_security_policy_script_nonce
    // my JavaScript
```

I'm currently not impacted with inline stylesheets (via `<style>`) but that should be the same problem.

[1] https://csp.withgoogle.com/docs/index.html
[2] https://github.com/twitter/secureheaders

**Comment by k0kubun** [comment]
> Looking at the current implementation it seems as if Haml::Filters::Javascript.render_with_options is not designed to take attributes. Is this intentional? I wasn't able to fully follow how #render_with_options is invoked and where it's options parameter comes from

See https://github.com/haml/haml/blob/v5.0.1/lib/haml/filters.rb#L180-L185.
It takes `Haml::Engine#options` or its subset `Haml::Buffer#options`. It doesn't suit your use case.

> Ideally I'd like to write something like this
>   :javascript noince: content_security_policy_script_nonce

I'm reluctant to introduce such syntax because the use case is very limited. Other possible attributes are configurable by `Haml::Engine#options`.

I recommend you to solve this problem with having special filters like:

```rb
module Haml
  module Filters
    module SecureJavascript
      include Base

      def render_with_options(*args)
        Javascript.render_with_options(*args).sub(/\A<script([^>]*)>/, "<script\\1 nonce='#{content_security_policy_script_nonce}'>")
      end
    end
  end
end

```

```haml
:securejavascript
  // your javascript
```

**Comment by tisba** [comment]
I'll give the custom filter a try. Thanks a lot!

**Comment by tisba** [comment]
I'm just now coming around to give this another try :-/

Anyway, @k0kubun, is there a way to access view helpers? The method `content_security_policy_script_nonce` is defined a a view helper method by the secure header gem.

**Comment by k0kubun** [comment]
> Anyway, @k0kubun, is there a way to access view helpers?

From haml filters, probably no.

So you have 2 options:

* Find some API in secure header gem that is available in haml filters
* Just don't use filter for this, render script tag like `%script{ nonce: content_security_policy_script_nonce }= js` 

Probably we can find another way if we read Haml internals carefully, but I don't have time to do that for now. Sorry.

**Comment by tisba** [comment]
The problem is, if I read the secureheader gem code correctly (and from what it does), that we need access to the request context since the nonce is per request.

Anyway, thanks again, maybe I'll go with the `%script` way for now. There aren't many inline scripts left over anyway :)

**Comment by jairovm** [comment]
I was able to add `nonce` attribute to my `<script>` and `<style>` tags by doing these two changes.

- Add a before_action filter in `ApplicationController` so `content_security_policy_nonce` value is saved using [ActiveSupport::CurrentAttributes](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html) so it can be accessed later from within `Haml` filters
  ```ruby
  # app/controllers/application_controller.rb

  before_action do
    Current.content_security_policy_nonce = content_security_policy_nonce
  end
  ```

- Override Haml filters
  ```ruby
  # config/initializers/haml_filters.rb

  Haml::Filters::Javascript.module_eval do
    alias_method :orig_render_with_options, :render_with_options

    def render_with_options(*args)
      orig_render_with_options(*args).sub(
        /\A<script([^>]*)>/, "<script\\1 nonce=\"#{Current.content_security_policy_nonce}\">"
      )
    end
  end

  Haml::Filters::Css.module_eval do
    alias_method :orig_render_with_options, :render_with_options

    def render_with_options(*args)
      orig_render_with_options(*args).sub(
        /\A<style([^>]*)>/, "<style\\1 nonce=\"#{Current.content_security_policy_nonce}\">"
      )
    end
  end
  ```

The good thing about this approach is that there's no need of creating new `filters`, `:javascript` and `:css` add `nonce` tag now.

**Comment by rassom** [comment]
@jairovm Tried your solution (with Rails 5.2.4.3), but get a n `uninitialized constant ApplicationController::Current`error. Any idea why?

**Comment by tbk303** [comment]
@jairovm Does this really work for you? I have set this up but the nonce in the nonce attribute of the script tag does not change from request to request if the script tag does not contain any Ruby variable interpolation. It seems like HAML is somehow caching those "static" script blocks.

**Comment by jairovm** [comment]
It's been working fine for me, make sure you uncomment this out in `content_security_policy.rb` initializer

```ruby
# If you are using UJS then enable automatic nonce generation
Rails.application.config.content_security_policy_nonce_generator = ->(_request) { SecureRandom.base64(16) }
```

**Comment by rassom** [comment]
@jairovm Thanks for the solution :) 

I don't understand "uncomment this out" in your last message.

Do you mean it should look like this:

````
# If you are using UJS then enable automatic nonce generation
# Rails.application.config.content_security_policy_nonce_generator = ->(_request) { SecureRandom.base64(16) }
````
or this:

````
# If you are using UJS then enable automatic nonce generation
Rails.application.config.content_security_policy_nonce_generator = ->(_request) { SecureRandom.base64(16) }
````

Thanks :)

**Comment by jairovm** [comment]
It should look like this
```ruby
# If you are using UJS then enable automatic nonce generation
Rails.application.config.content_security_policy_nonce_generator = ->(_request) { SecureRandom.base64(16) }
```

👍🏼

**Comment by rassom** [comment]
@jairovm 

It has worked like a charm until today where I updated my Haml gems:

````
Installing haml 6.0.1 (was 5.2.2) with native extensions
Installing haml-rails 2.1.0 (was 2.0.1)
````

Now I get this error:
````
config/initializers/haml_filters.rb:2:in `alias_method': undefined method `render_with_options' for class `Haml::Filters::Javascript' (NameError)

  alias_method :orig_render_with_options, :render_with_options
  ^^^^^^^^^^^^
````

Any ideas how to fix that?

**Comment by jairovm** [comment]
I'm still using haml 5 😅 sorry, I'll take a look when I have a chance.

**Comment by rassom** [comment]
Thank you! :)

**Comment by rassom** [comment]
@jairovm Did you get a chance to take a look? 🥳

**Comment by dingjiedanielyang-sec** [comment]
Thank you @jairovm for the detailed solution.  @rassom, I just overwrite the :render_with_options function entirely with the following code and it works as expected
`Haml::Filters::Javascript.module_eval do
  def render_with_options(text, options)
    indent = options[:cdata] ? '    ' : '  ' # 4 or 2 spaces
    if options[:format] == :html5
      type = ''
    else
      type = " type=#{options[:attr_wrapper]}text/javascript#{options[:attr_wrapper]}"
    end

    text = text.rstrip
    text.gsub!("\n", "\n#{indent}")

    %(<script#{type} nonce="#{Current.content_security_policy_nonce}">\n#{"  //<![CDATA[\n" if options[:cdata]}#{indent}#{text}\n#{"  //]]>\n" if options[:cdata]}</script>)
  end

end
`

**Comment by rassom** [comment]
@dingjiedanielyang-sec  Thank you.

Would you please help with the CSS part also? 🙏

````
Haml::Filters::Css.module_eval do
  alias_method :orig_render_with_options, :render_with_options

  def render_with_options(*args)
    orig_render_with_options(*args).sub(
      /\A<style([^>]*)>/, "<style\\1 nonce=\"#{Current.content_security_policy_nonce}\">"
    )
  end
end
````

**Comment by dingjiedanielyang-sec** [comment]
Not a problem. @rassom Could you please try the following part for the CSS? 
```
Haml::Filters::Css.module_eval do
      # @see Base#render_with_options
      def render_with_options(text, options)
        indent = options[:cdata] ? '    ' : '  ' # 4 or 2 spaces
        if options[:format] == :html5
          type = ''
        else
          type = " type=#{options[:attr_wrapper]}text/css#{options[:attr_wrapper]}"
        end

        text = text.rstrip
        text.gsub!("\n", "\n#{indent}")

        %(<style#{type} nonce="#{Current.content_security_policy_nonce}">\n#{"  /*<![CDATA[*/\n" if options[:cdata]}#{indent}#{text}\n#{"  /*]]>*/\n" if options[:cdata]}</style>)
      end
  
  end
```

**Comment by Morozzzko** [comment]
So I struggled with this issue @ Haml 5. First Haml 5 solutions _would not work_ unless you have interpolation in your scripts.
This happens because Haml compiles the templates. If the script is static, there's no reason to re-evaluate it. And the patches did indeed generate static scripts.

However, Haml 6 has a better & a more solid solution. We can just patch `Haml::Filters::Javascript` and override `render_html` method to return a _dynamic_ script

See original source file, pointing to the line we'll need to modify:

https://github.com/haml/haml/blob/d0a86c3eba4563073e950f9e7307a947d4379a64/lib/haml/filters/javascript.rb#L18

Basically, we can replace this with a custom [Temple](https://github.com/judofyr/temple/blob/master/EXPRESSIONS.md) expression which would calculate the values dynamically

Here's what I got:

```ruby
    script = <<~RUBY
      %(<script nonce="\#{content_security_policy_nonce}">\n)
    RUBY
    temple << [:dynamic, script]
```

What you see here is that it _does not require_ `Current`. It integrates with view helpers and requires fewer lines of code.

Hope it helps anyone

**Comment by rassom** [comment]
@Morozzzko Nice solution 👍  Do you have a similar solution for CSS also?

**Comment by Morozzzko** [comment]
@rassom I didn't really need it so, not really, but it should be almost identical, with `script` substituted for `style`

Here's the starting point for anyone curious:

https://github.com/haml/haml/blob/d0a86c3eba4563073e950f9e7307a947d4379a64/lib/haml/filters/css.rb#L18

**Comment by emptyflask** [comment]
This is an old thread, but I've used this on rails:
```haml
= javascript_tag nonce: true do
  :plain
    console.log('it works')
```

**Comment by arianf** [comment]
It would be great if haml supported a solution for this natively

**Comment by arianf** [comment]
I went with something like this: `config/initializers/haml_csp_nonce_support.rb`

```ruby
require 'haml'

# Version compatibility check
if !Gem::Requirement.new('~> 6.3').satisfied_by?(Gem::Version.new(Haml::VERSION))
  raise "Haml CSP nonce support hasn't been tested with versions. " \
        'Please update this initializer or test compatibility.'
end

# https://github.com/haml/haml/blob/v6.3.0/lib/haml/filters/javascript.rb
# https://github.com/haml/haml/blob/v6.3.0/lib/haml/filters/css.rb

# Content Security Policy (CSP) nonce integration for Haml filters
# This modification adds automatic nonce attributes to inline scripts and styles
# to comply with CSP security policies while maintaining functionality
module Haml
  class Filters
    # Enhanced Javascript filter with CSP nonce support
    class Javascript < TextBase
      def compile_html(node)
        temple = [:multi]

        # === ORIGINAL CODE ===
        # temple << [:static, "<script>\n"]

        # === CSP ENHANCEMENT START ===
        temple << [:static,  '<script nonce="']
        temple << [:dynamic, 'content_security_policy_nonce']
        temple << [:static,  "\">\n"]
        # === CSP ENHANCEMENT END ===

        compile_text!(temple, node, '  ')
        temple << [:static, "\n</script>"]
        temple
      end
    end

    # Enhanced CSS filter with CSP nonce support
    class Css < TextBase
      def compile_html(node)
        temple = [:multi]

        # === ORIGINAL CODE ===
        # temple << [:static, "<style>\n"]

        # === CSP ENHANCEMENT START ===
        temple << [:static,  '<style nonce="']
        temple << [:dynamic, 'content_security_policy_nonce']
        temple << [:static,  "\">\n"]
        # === CSP ENHANCEMENT END ===

        compile_text!(temple, node, '  ')
        temple << [:static, "\n</style>"]
        temple
      end
    end
  end
end
```
